### PR TITLE
quantlib: add version 1.30

### PR DESCRIPTION
**Pull request by BrettDong**
Specify library name and version:  **quantlib/1.30**

---

- [X] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [X] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [X] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by RubenRBS** [comment]
You have a failing Hook in the v1 pipeline:
````
[HOOK - conan-center.py] post_package(): ERROR: [CMAKE-MODULES-CONFIG-FILES (KB-H016)]
The conan-center repository doesn't allow the packages to contain CMake find modules or config files.
The packages have to be located using generators and the declared `cpp_info` information 
(https://github.com/conan-io/conan-center-index/blob/master/docs/error_knowledge_base.md#KB-H016-CMAKE-MODULES-CONFIG-FILES)```

**Review by memsharded — APPROVED** [review]
Looking good, thanks for your contribution!

**Review by RubenRBS — APPROVED** [review]

File: recipes/quantlib/all/conanfile.py [hunk 1]
```diff
@@ -43,27 +43,38 @@ def layout(self):
...
+        self.requires("boost/1.80.0", transitive_headers=True)
 
     def validate(self):
         if self.info.settings.compiler.get_safe("cppstd"):
-            check_min_cppstd(self, 11)
+            check_min_cppstd(self, 14 if self.version >= "1.24" else 11)
         if self.info.settings.compiler == "gcc" and Version(self.info.settings.compiler.version) < "5":
             raise ConanInvalidConfiguration("gcc < 5 not supported")
+        if self.version >= "1.24" and is_msvc(self) and self.options.shared:
+            raise ConanInvalidConfiguration("MSVC DLL build is not supported by upstream")
```
Comment (SpaceIm) [inline recipes/quantlib/all/conanfile.py:54]: I don't understand, is CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS not sufficient? Why is it not injected for this specific version?

Comment (BrettDong) [reply recipes/quantlib/all/conanfile.py:54]: I don't know why upstream does not allow shared build with MSVC, and I don't have a Windows device to try, so I just leave it consistent with upstream instead of a patch.

File: recipes/quantlib/all/conanfile.py [hunk 2]
```diff
@@ -85,6 +96,8 @@ def package_info(self):
         self.cpp_info.set_property("pkg_config_name", "quantlib")
         self.cpp_info.libs = collect_libs(self)
         self.cpp_info.requires = ["boost::headers"]
+        if self.settings.os in ["Linux", "FreeBSD"] and self.options.shared:
+            self.cpp_info.system_libs.append("m")
```
[the remark this answers is not in the document]
Comment (BrettDong) [reply recipes/quantlib/all/conanfile.py:100]: Yes I added this after seeing the hook warning. Should I remove this block?

File: recipes/quantlib/all/conanfile.py (hunk 1 above)
Comment (SpaceIm) [reply recipes/quantlib/all/conanfile.py:54]: It makes recipe inconsistent. I don't understand your answer. Is it prevented or is it broken by default without `CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS` like previous version.

File: recipes/quantlib/all/conanfile.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (SpaceIm) [reply recipes/quantlib/all/conanfile.py:100]: No, it should be added if static, not shared.

File: recipes/quantlib/all/conanfile.py (hunk 1 above)
Comment (BrettDong) [reply recipes/quantlib/all/conanfile.py:54]: It's explicitly prevented in the upstream CMake script.

Comment (SpaceIm) [reply recipes/quantlib/all/conanfile.py:54]: It's patched in 1.22, I don't see any reason to not do the same in 1.30 since it works just fine and it's a one line patch.

Comment (BrettDong) [reply recipes/quantlib/all/conanfile.py:54]: You can add it back in a subsequent pull request. I don't have a Windows device to test if it works so I'd better be conservative in this pull request.
